sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,718 @@
+
+FAST_METRICS = KDTree.valid_metrics + BallTree.valid_metrics
+
+
+def _tree_to_labels(
+    single_linkage_tree,
+    min_cluster_size=10,
+    cluster_selection_method="eom",
+    allow_single_cluster=False,
+    cluster_selection_epsilon=0.0,
+    max_cluster_size=0,
+):
+    """Converts a pretrained tree and cluster size into a
+    set of labels and probabilities.
+    """
+    condensed_tree = condense_tree(single_linkage_tree, min_cluster_size)
+    stability_dict = compute_stability(condensed_tree)
+    labels, probabilities = get_clusters(
+        condensed_tree,
+        stability_dict,
+        cluster_selection_method,
+        allow_single_cluster,
+        cluster_selection_epsilon,
+        max_cluster_size,
+    )
+
+    return (labels, probabilities, single_linkage_tree)
+
+
+def _process_mst(min_spanning_tree):
+    # Sort edges of the min_spanning_tree by weight
+    row_order = np.argsort(min_spanning_tree.T[2])
+    min_spanning_tree = min_spanning_tree[row_order, :]
+    # Convert edge list into standard hierarchical clustering format
+    return label(min_spanning_tree)
+
+
+def _hdbscan_brute(
+    X,
+    min_samples=5,
+    alpha=1.0,
+    metric="euclidean",
+    n_jobs=None,
+    **metric_params,
+):
+    if metric == "precomputed":
+        # Treating this case explicitly, instead of letting
+        # sklearn.metrics.pairwise_distances handle it,
+        # enables the usage of numpy.inf in the distance
+        # matrix to indicate missing distance information.
+        distance_matrix = X
+    else:
+        distance_matrix = pairwise_distances(
+            X, metric=metric, n_jobs=n_jobs, **metric_params
+        )
+
+    if issparse(distance_matrix):
+        return _hdbscan_sparse_distance_matrix(
+            distance_matrix,
+            min_samples,
+            alpha,
+            **metric_params,
+        )
+
+    mutual_reachability_ = mutual_reachability(distance_matrix, min_samples, alpha)
+
+    min_spanning_tree = mst_linkage_core(mutual_reachability_)
+
+    # Warn if the MST couldn't be constructed around the missing distances
+    if np.isinf(min_spanning_tree.T[2]).any():
+        warn(
+            "The minimum spanning tree contains edge weights with value "
+            "infinity. Potentially, you are missing too many distances "
+            "in the initial distance matrix for the given neighborhood "
+            "size.",
+            UserWarning,
+        )
+
+    return _process_mst(min_spanning_tree)
+
+
+def _hdbscan_sparse_distance_matrix(
+    X,
+    min_samples=5,
+    alpha=1.0,
+    **metric_params,
+):
+    # Compute sparse mutual reachability graph
+    # if max_dist > 0, max distance to use when the reachability is infinite
+    max_dist = metric_params.get("max_dist", 0.0)
+    mutual_reachability_ = sparse_mutual_reachability(
+        X.tolil(), min_points=min_samples, max_dist=max_dist, alpha=alpha
```

Comment:
In the short-term, I've added `LIL` to the list of preservable sparse formats in validation. Long-term, rewriting the Cython routine to operate directly on `CSR` is the best move. I currently have that as a follow-up task, since without that it still works (albeit with a rather obvious cost).